File: Python/demo_gtmmlr.py
# -*- coding: utf-8 -*- 
#%reset -f
"""
@author: Hiromasa Kaneko
"""

# Demonstration of GTM-MLR (Generative Topographic Mapping - Multiple Linear Regression)

# settings
shapeofmap = [30, 30]
shapeofrbfcenters = [4, 4]
varianceofrbfs = 0.5
lambdainemalgorithm = 0.001
numberofiterations = 300
desplayflag = 1
numofsamples = 1000
noisey = 0.1
random_state_number = 30000

import numpy as np
#import pandas as pd
from sklearn.datasets.samples_generator import make_swiss_roll
from gtm import gtm
import matplotlib.pyplot as plt
import matplotlib.figure as figure
from mpl_toolkits.mplot3d import Axes3D

# load a swiss roll dataset and make a y-variable
OriginalX, color = make_swiss_roll(numofsamples, 0, random_state=10)
X = OriginalX
rawy = 0.3 * OriginalX[:,0] - 0.1 * OriginalX[:,1] + 0.2 * OriginalX[:,2]
originaly = rawy + noisey * rawy.std(ddof=1) * np.random.randn(len(rawy))
# plot
plt.rcParams["font.size"] = 18
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
p = ax.scatter(OriginalX[:,0], OriginalX[:,1], OriginalX[:,2], c=originaly)
fig.colorbar(p)
plt.show()

# divide a dataset into training data and test data
Xtrain = OriginalX[:500,:]
ytrain = originaly[:500]
Xtest = OriginalX[500:,:]
ytest = originaly[500:]

# autoscaling
#autoscaledX = (OriginalX - OriginalX.mean(axis=0)) / OriginalX.std(axis=0,ddof=1)
autoscaledXtrain = (Xtrain - Xtrain.mean(axis=0)) / Xtrain.std(axis=0,ddof=1)
#autoscaledXtest = (Xtest - X.mean(axis=0)) / X.std(axis=0,ddof=1)
#autoscaledytrain = (ytrain - ytrain.mean()) / ytrain.std(ddof=1)

# construct GTM model
model = gtm( shapeofmap, shapeofrbfcenters, varianceofrbfs, lambdainemalgorithm, numberofiterations, desplayflag)
model.fit(autoscaledXtrain)
if model.successflag:
    # calculate of responsibilities
    responsibilities = model.responsibility(autoscaledXtrain)
    
    # plot the mean of responsibilities
    means = responsibilities.dot( model.mapgrids )
    plt.figure()
#    plt.figure(figsize=figure.figaspect(1))
    plt.scatter( means[:,0], means[:,1], c=ytrain)
    plt.colorbar()
    plt.ylim(-1.1,1.1)
    plt.xlim(-1.1,1.1)
    plt.xlabel("z1 (mean)")
    plt.ylabel("z2 (mean)")
    plt.show()
    
    # plot the mode of responsibilities
    modes = model.mapgrids[responsibilities.argmax(axis=1), :]
    plt.figure()
#    plt.figure(figsize=figure.figaspect(1))
    plt.scatter( modes[:,0], modes[:,1], c=ytrain)
    plt.colorbar()
    plt.ylim(-1.1,1.1)
    plt.xlim(-1.1,1.1)
    plt.xlabel("z1 (mode)")
    plt.ylabel("z2 (mode)")
    plt.show()

# construct MLR model
model.mlr(Xtrain,ytrain)

# MLR prediction
predictedytest = np.ndarray.flatten( model.mlrpredict(Xtest) )

# r2p, RMSEp, MAEp
print( "r2p: {0}".format(float( 1 - sum( (ytest-predictedytest )**2 ) / sum((ytest-ytest.mean())**2) )) )
print( "RMSEp: {0}".format(float( ( sum( (ytest-predictedytest)**2 )/ len(ytest))**(1/2) )) )
print( "MAEp: {0}".format(float( sum( abs(ytest-predictedytest)) / len(ytest) )) )
# yyplot
plt.figure(figsize=figure.figaspect(1))
plt.scatter( ytest, predictedytest)
YMax = np.max( np.array([np.array(ytest), predictedytest]))
YMin = np.min( np.array([np.array(ytest), predictedytest]))
plt.plot([YMin-0.05*(YMax-YMin),YMax+0.05*(YMax-YMin)], [YMin-0.05*(YMax-YMin),YMax+0.05*(YMax-YMin)], 'k-')
plt.ylim(YMin-0.05*(YMax-YMin),YMax+0.05*(YMax-YMin))
plt.xlim(YMin-0.05*(YMax-YMin),YMax+0.05*(YMax-YMin))
plt.xlabel("simulated y")
plt.ylabel("estimated y")
plt.show()

File: MATLAB/demo_GTMMLR.m
clear; close all;
% Demonstration of GTM-MLR (Generative Topographic Mapping - Multiple Linear Regression) using a swiss roll dataset

targetyvalue = 4; % y-target for inverse analysis

% settings
shapeofmap = [ 30, 30];
shapeofrbfcenters = [ 4, 4];
varianceofrbfs = 0.5;
lambdainemalgorithm = 0.001;
numberofiterations = 300;
displayflag = 1;
k = 10;

% load a dataset
dataset = csvread('swissroll.csv');
y = dataset(:,3);
X = dataset(:,4:end);

figure;
scatter3(X(:,1),X(:,2),X(:,3),[],y,'filled');
colormap(jet);
colorbar;
xlim([ min(X(:,1))-range(X(:,1))*0.03 max(X(:,1))+range(X(:,1))*0.03] );
ylim([ min(X(:,2))-range(X(:,2))*0.03 max(X(:,2))+range(X(:,2))*0.03] );
zlim([ min(X(:,3))-range(X(:,3))*0.03 max(X(:,3))+range(X(:,3))*0.03] );
axis square;
xlabel( 'x_1' ,  'FontSize' , 18 , 'FontName', 'Times');
ylabel( 'x_2' ,  'FontSize' , 18 , 'FontName', 'Times');
zlabel( 'x_3' ,  'FontSize' , 18 , 'FontName', 'Times');
set(gcf, 'Color' , 'w' ); 
set(gca, 'FontSize', 18);
set(gca, 'FontName', 'Times');

% divide a dataset into training data and test data
traindatanumber = 500;
ytrain = y(1:traindatanumber,:); Xtrain = X(1:traindatanumber,:);
ytest = y(traindatanumber+1:end,:); Xtest = X(traindatanumber+1:end,:);

% autoscaling
autoscaledXtrain = ( Xtrain - repmat(mean(Xtrain),size(Xtrain,1),1) ) ./ repmat(std(Xtrain),size(Xtrain,1),1);
autoscaledytrain = ( ytrain - mean(ytrain) ) ./ std(ytrain);
autoscaledXtest = ( Xtest - repmat(mean(Xtrain),size(Xtest,1),1) ) ./ repmat(std(Xtrain),size(Xtest,1),1);

% construct GTM model
gtmmodel = calc_gtm(autoscaledXtrain, shapeofmap,shapeofrbfcenters, varianceofrbfs, lambdainemalgorithm, numberofiterations, 0);
if gtmmodel.successflag
    % calculate responsibility
    responsibilities = calc_responsibility(gtmmodel, autoscaledXtrain);
    % plot the mean of the responsibility
    xpred_mean = responsibilities * gtmmodel.mapgrids;
    figure; scatter( xpred_mean(:,1), xpred_mean(:,2), [], ytrain, 'filled');
    colormap(jet); colorbar;
    axis square; 
    xlabel( 'z_1 (mean)' ,  'FontSize' , 18 , 'FontName','Times');
    ylabel( 'z_2 (mean)' ,  'FontSize' , 18 , 'FontName','Times');
    set(gcf, 'Color' , 'w' ); 
    set(gca, 'FontSize' , 18);
    set(gca, 'FontName','Times');
    axis([-1.1 1.1 -1.1 1.1]);
    % plot the mode of the responsibility
    [~, maxindex] = max(responsibilities, [], 2);
    xpred_mode = gtmmodel.mapgrids(maxindex, :);
    figure; scatter( xpred_mode(:,1), xpred_mode(:,2), [], ytrain, 'filled');
    colormap(jet); colorbar;
    axis square;
    xlabel( 'z_1 (mode)' ,  'FontSize' , 18 , 'FontName','Times');
    ylabel( 'z_2 (mode)' ,  'FontSize' , 18 , 'FontName','Times');
    set(gcf, 'Color' , 'w' ); 
    set(gca, 'FontSize' , 18);
    set(gca, 'FontName','Times');
    axis([-1.1 1.1 -1.1 1.1]);
end
% construct MLR model
[calculatedy, regressioncoefficients] = mlr_calc(autoscaledXtrain, autoscaledytrain, autoscaledXtrain);
calculatedy = calculatedy * std(ytrain) + mean(ytrain);
sigma = 1/length(y) * sum( (ytrain - calculatedy).^2 );
% MLR prediction
ytestpred = autoscaledXtest * regressioncoefficients;
ytestpred = ytestpred * std(ytrain) + mean(ytrain);
figure;
plot( ytest, ytestpred, 'b.', 'MarkerSize' , 15 );
hold on;
plot( [-6.5 5.5] , [-6.5 5.5] , 'k' , 'LineWidth' , 2 );
xlabel('simulated y' , 'FontSize' , 18 , 'FontName', 'Times');
ylabel('estimated y' , 'FontSize' , 18 , 'FontName', 'Times');
axis( [ -6.5 5.5 -6.5 5.5 ] );
hold off;
set(gcf, 'Color' , 'w' );
set(gca, 'FontSize' ,20 );
axis square;
r2p = 1 - ( ( ytest - ytestpred )' * ( ytest - ytestpred ) ) ./ ( ( ytest - mean(ytest) )' * ( ytest - mean(ytest) ) );
maep = sum( abs( ytest - ytestpred ) ) / length(ytest);
disp( 'r2p and MAEp' );
disp( [ r2p, maep] );
